src/platforms/stm32/src/lib/gpio_driver.c

```diff
@@ -229,6 +229,22 @@ static term create_pair(Context *ctx, term term1, term term2)
     return ret;
 }
 
+// term `error` can be an atom or a `const char *const ATOM_STR`
+static term error_tuple_maybe_gc(Context *ctx, term error)
+{
+    term error_atom;
+    if (!term_is_atom(error)) {
+        error_atom = globalcontext_make_atom(ctx->global, (const char *const) error);
+    } else {
+        error_atom = error;
+    }
+    if (UNLIKELY(memory_ensure_free(ctx, TUPLE_SIZE(2)) != MEMORY_GC_OK)) {
+        return OUT_OF_MEMORY_ATOM;
+    }
+
+    return create_pair(ctx, ERROR_ATOM, error_atom);
+}
```

Comment:
Another (safer) option is adding a new atom string specialized function, that inserts the atom and then calls the original function.

Reply:
I like this option better. I added `error_tuple_str_maybe_gc` to accept atom stings, which will create the `Reason` atom and pass it to `error_tuple_maybe_gc`.